test/Dialect/Arc/lower-to-llvm.mlir

```diff
@@ -0,0 +1,151 @@
+
+// CHECK-LABEL: llvm.func @StateUpdates(%arg0: !llvm.ptr<i8>) {
+func.func @StateUpdates(%arg0: !arc.storage<1>) {
+  %0 = arc.alloc_state %arg0 {offset = 0} : (!arc.storage<1>) -> !arc.state<i1>
+  // CHECK-NEXT: [[OFFSET:%.+]] = llvm.mlir.constant(0 :
+  // CHECK-NEXT: [[RAW_PTR:%.+]] = llvm.getelementptr %arg0[[[OFFSET]]]
+  // CHECK-NEXT: [[PTR:%.+]] = llvm.bitcast [[RAW_PTR]] : !llvm.ptr<i8> to !llvm.ptr<i1>
+
+  %1 = arc.state_read %0 : <i1>
+  // CHECK-NEXT: [[TMP:%.+]] = llvm.load [[PTR]]
+  arc.state_write %0 = %1 : <i1>
+  // CHECK-NEXT: llvm.store [[TMP]], [[PTR]]
+  return
+  // CHECK-NEXT: llvm.return
+}
+// CHECK-NEXT: }
+
+// CHECK-LABEL: llvm.func @MemoryUpdates(%arg0: !llvm.ptr<i8>, %arg1: i1) {
+func.func @MemoryUpdates(%arg0: !arc.storage<24>, %enable: i1) {
+  %0 = arc.alloc_memory %arg0 {offset = 0, stride = 6} : (!arc.storage<24>) -> !arc.memory<4 x i42, 6>
+  // CHECK-NEXT: [[OFFSET:%.+]] = llvm.mlir.constant(0 :
+  // CHECK-NEXT: [[RAW_PTR:%.+]] = llvm.getelementptr %arg0[[[OFFSET]]]
+  // CHECK-NEXT: [[PTR:%.+]] = llvm.bitcast [[RAW_PTR]] : !llvm.ptr<i8> to !llvm.ptr<i48>
+
+  %clk = hw.constant true
+  %c3_i19 = hw.constant 3 : i19
+  // CHECK-NEXT: llvm.mlir.constant(true
+  // CHECK-NEXT: [[THREE:%.+]] = llvm.mlir.constant(3
+
+  %1 = arc.memory_read %0[%c3_i19], %clk, %enable : <4 x i42, 6>, i19
+  %2 = comb.add %1, %1 : i42
+  // CHECK-NEXT:   [[ADDR:%.+]] = llvm.zext [[THREE]] : i19 to i20
+  // CHECK-NEXT:   [[FOUR:%.+]] = llvm.mlir.constant(4
+  // CHECK-NEXT:   [[INBOUNDS:%.+]] = llvm.icmp "ult" [[ADDR]], [[FOUR]]
+  // CHECK-NEXT:   [[GEP:%.+]] = llvm.getelementptr [[PTR]][[[ADDR]]] : (!llvm.ptr<i48>, i20) -> !llvm.ptr<i42>
+  // CHECK-NEXT:   [[COND:%.+]] = llvm.and %arg1, [[INBOUNDS]]
+  // CHECK-NEXT:   llvm.cond_br [[COND]], [[BB_LOAD:\^.+]], [[BB_SKIP:\^.+]]
+  // CHECK-NEXT: [[BB_LOAD]]:
+  // CHECK-NEXT:   [[TMP:%.+]] = llvm.load [[GEP]]
+  // CHECK-NEXT:   llvm.br [[BB_RESUME:\^.+]]([[TMP]] : i42)
+  // CHECK-NEXT: [[BB_SKIP]]:
+  // CHECK-NEXT:   [[TMP:%.+]] = llvm.mlir.constant
+  // CHECK-NEXT:   llvm.br [[BB_RESUME:\^.+]]([[TMP]] : i42)
+  // CHECK-NEXT: [[BB_RESUME]]([[LOADED:%.+]]: i42):
+  // CHECK:        [[ADDED:%.+]] = llvm.add [[LOADED]], [[LOADED]]
+
+  arc.memory_write %0[%c3_i19], %clk, %enable, %2 : <4 x i42, 6>, i19
+  // CHECK-NEXT:   [[ADDR:%.+]] = llvm.zext [[THREE]] : i19 to i20
+  // CHECK-NEXT:   [[FOUR:%.+]] = llvm.mlir.constant(4
+  // CHECK-NEXT:   [[INBOUNDS:%.+]] = llvm.icmp "ult" [[ADDR]], [[FOUR]]
+  // CHECK-NEXT:   [[GEP:%.+]] = llvm.getelementptr [[PTR]][[[ADDR]]] : (!llvm.ptr<i48>, i20) -> !llvm.ptr<i42>
+  // CHECK-NEXT:   [[COND:%.+]] = llvm.and %arg1, [[INBOUNDS]]
+  // CHECK-NEXT:   llvm.cond_br [[COND]], [[BB_STORE:\^.+]], [[BB_RESUME:\^.+]]
+  // CHECK-NEXT: [[BB_STORE]]:
+  // CHECK-NEXT:   llvm.store [[ADDED]], [[GEP]]
+  // CHECK-NEXT:   llvm.br [[BB_RESUME]]
+  // CHECK-NEXT: [[BB_RESUME]]:
+  return
+  // CHECK-NEXT:   llvm.return
+}
+// CHECK-NEXT: }
```

Comment:
ClockTree and PassThrough lowerings are not tested, but since there is the separate clock tree lowering pass, they are not in use anymore. Should we keep them and test them, or remove them?

Reply:
Removed them